### DISCUSSION TITLE
`in_` `notin_` no longer works on results from column-specific queries

### DISCUSSION THREAD

**Discussion by ljiatu** [sqlalchemy/sqlalchemy#6117] [Usage Questions] [answered]
**Describe the bug**

Not quite sure how I should summarize the bug, but basically the code below used to work in 1.3.* but not anymore in 1.4.2. Doesn't seem like this is an intended breaking change?

**To Reproduce**
```py
@as_declarative
class User():
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)

def test_sqlalchemy(db_session: Session) -> None:
    db_users = [User(id=uuid.uuid4()), User(id=uuid.uuid4()), User(id=uuid.uuid4())]
    db_session.add_all(db_users)
    db_session.commit()

    user_ids = db_session.query(User.id).all()
    db_session.query(User).filter(User.id.in_(user_ids)).all()
```

**Error**

```
self = <sqlalchemy.dialects.postgresql.psycopg2.PGDialect_psycopg2 object at 0x10e2be430>, cursor = <cursor object at 0x10e19ac70; closed: -1>
statement = 'SELECT users.id AS users_id, users.created_at AS users_created_at, users.updated_at AS users_updated_at, users.email ..., users.email_confirmed AS users_email_confirmed \nFROM users \nWHERE users.id IN (%(id_1_1)s, %(id_1_2)s, %(id_1_3)s)'
parameters = {'id_1_1': (UUID('3bd1f4a2-7092-4825-9271-4ff63c315fd7'),), 'id_1_2': (UUID('7e1b20d9-18b1-4fc8-b8bb-45517079f17d'),), 'id_1_3': (UUID('e2296df7-507d-4e4f-bd9f-d665cfe63537'),)}
context = <sqlalchemy.dialects.postgresql.psycopg2.PGExecutionContext_psycopg2 object at 0x10e5d9a00>

    def do_execute(self, cursor, statement, parameters, context=None):
>       cursor.execute(statement, parameters)
E       sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) can't adapt type 'Row'
E       [SQL: SELECT users.id AS users_id, users.created_at AS users_created_at, users.updated_at AS users_updated_at, users.email AS users_email, users.password AS users_password, users.first_name AS users_first_name, users.last_name AS users_last_name, users.password_reset_at AS users_password_reset_at, users.locked_at AS users_locked_at, users.current_failed_sign_in_attempt AS users_current_failed_sign_in_attempt, users.email_confirmed AS users_email_confirmed
E       FROM users
E       WHERE users.id IN (%(id_1_1)s, %(id_1_2)s, %(id_1_3)s)]
E       [parameters: {'id_1_1': (UUID('3bd1f4a2-7092-4825-9271-4ff63c315fd7'),), 'id_1_2': (UUID('7e1b20d9-18b1-4fc8-b8bb-45517079f17d'),), 'id_1_3': (UUID('e2296df7-507d-4e4f-bd9f-d665cfe63537'),)}]
E       (Background on this error at: http://sqlalche.me/e/14/f405)
```

**Versions.**
 - OS: maxOS Catalina 10.15.6 (it's not OS specific)
 - Python: 3.8
 - SQLAlchemy: 1.4.2
 - Database: Postgres 12
 - DBAPI: psycopg2

**Have a nice day!**

**Comment by zzzeek** [comment, marked as answer]
hi -

thanks for the complete test case.   I don't see a bug here because the issue is that you are passing a list of Row() instances to the in_() operator, which is not unlike passing a list of tuple() instances, which in both cases is not stating your actual intent, which is to pass a list of uuid() objects.  In 1.3, you are passing a series of KeyedTuple objects instead of Row; for background on this change see https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#the-keyedtuple-object-returned-by-query-is-replaced-by-row .

For whatever reason, psycopg2 seems fine if we send the UUID parameter inside of a tuple or inside of 1.3's KeyedTuple, here's the 1.3 log:

```

SELECT "user".id AS user_id 
FROM "user" 
WHERE "user".id IN (%(id_1)s, %(id_2)s, %(id_3)s)
2021-03-23 10:35:08,485 INFO sqlalchemy.engine.base.Engine {'id_1': (UUID('f02d02c7-c9d4-41e7-9c42-8709fff3c933'),), 'id_2': (UUID('10b9b69d-9cdf-44b0-aa30-2ef1d649a1b5'),), 'id_3': (UUID('b7ccf410-06d9-4ea4-bbb9-0edc2e86c925'),)}
```

note how each bound value is inside of a tuple, like `(UUID('10b9b69d-9cdf-44b0-aa30-2ef1d649a1b5'),)`.  I don't know why psycopg2 accepts that and somehow knows to get the first element of the tuple.

Given that we want to pass a list of uuids to the operator, and not a list of tuples of uuids, the code that works in both 1.3 and 1.4 in this case is:

```python

user_ids = [u for u, in s.query(User.id).all()]
s.query(User).filter(User.id.in_(user_ids)).all()
```

using [1.4/2.0 style queries](https://docs.sqlalchemy.org/en/14/orm/queryguide.html) we can make this clearer:

```python
stmt = select(User.id)

user_ids = s.execute(stmt).scalars().all()

s.execute(select(User).filter(User.id.in_(user_ids))).all()

```

**Reply by ljiatu** [reply]
Ah ok. It worked in 1.3 so I assumed it was a bug. Thanks for the explanation!
